### PR TITLE
[SYCL][SCLA] Add `sycl::aspect::ext_oneapi_private_alloca`

### PR DESCRIPTION
**Pull request by victor-eds**
Add aspect to check for `sycl_ext_oneapi_private_alloca` extension support. Only SPIR-V (OpenCL and Level Zero) devices have this aspect.

In CodeGen, propagate this aspect when creating the `llvm.sycl.alloca.*` intrinsic.

In `sycl-post-link`, do not assert on unsupported targets. Instead, assume the default value for the size specialization constant is used and handle the error at runtime. SPIR-V AOT compilation should fail in the frontend, so no need to fail here.

### CODE REVIEW

File: llvm/test/tools/sycl-post-link/spec-constants/SYCL-alloca.ll [hunk 1]
```diff
@@ -30,10 +33,19 @@
...
+; CHECK-EMULATION:  alloca double, i32 120, align 8
   call ptr @llvm.sycl.alloca.p0.p4.p4.p4.f64(ptr addrspace(4) addrspacecast (ptr @size_i32_stable_name to ptr addrspace(4)), ptr addrspace(4) addrspacecast (ptr addrspace(1) @size_i32 to ptr addrspace(4)), ptr addrspace(4) null, double 0.000000e+00, i64 8)
 ; CHECK-RT: [[LENGTH:%.*]] = call i64 @_Z20__spirv_SpecConstantix(i32 0, i64 10)
 ; CHECK-RT: {{.*}} = alloca float, i64 [[LENGTH]], align 8
+
+; CHECK-EMULATION:  alloca float, i64 10, align 8
   call ptr @llvm.sycl.alloca.p0.p4.p4.p4.f32(ptr addrspace(4) addrspacecast (ptr @size_i64_stable_name to ptr addrspace(4)), ptr addrspace(4) addrspacecast (ptr addrspace(1) @size_i64 to ptr addrspace(4)), ptr addrspace(4) null, float 0.000000e+00, i64 8)
+
+; CHECK-RT:  %[[LENGTH:.*]] = call i16 @_Z20__spirv_SpecConstantis(i32 2, i16 1)
+; CHECK-RT:  {{.*}} = alloca %my_range, i16 %[[LENGTH]], align 64
```
Comment (victor-eds) [inline llvm/test/tools/sycl-post-link/spec-constants/SYCL-alloca.ll:46]: Previously missing

File: llvm/test/tools/sycl-post-link/spec-constants/SYCL-alloca.ll [hunk 2]
```diff
@@ -10,9 +13,9 @@
 %"class.sycl::_V1::specialization_id.1" = type { i16 }
 %my_range = type { ptr addrspace(4), ptr addrspace(4) }
 
-@size_i64 = internal addrspace(1) constant %"class.sycl::_V1::specialization_id" { i64 10 }, align 8
-@size_i32 = internal addrspace(1) constant %"class.sycl::_V1::specialization_id.0" { i32 120 }, align 4
-@size_i16 = internal addrspace(1) constant %"class.sycl::_V1::specialization_id.1" { i16 1 }, align 2
+@size_i64 = addrspace(1) constant %"class.sycl::_V1::specialization_id" { i64 10 }, align 8
+@size_i32 = addrspace(1) constant %"class.sycl::_V1::specialization_id.0" { i32 120 }, align 4
+@size_i16 = addrspace(1) constant %"class.sycl::_V1::specialization_id.1" { i16 1 }, align 2
```
Comment (victor-eds) [inline llvm/test/tools/sycl-post-link/spec-constants/SYCL-alloca.ll:18]: Needed to not drop

File: sycl/test-e2e/PrivateAlloca/ValidUsage/Inputs/private_alloca_test.hpp [hunk 3]
```diff
@@ -2,9 +2,10 @@
 
 // Template for private alloca tests.
 
-#include <sycl/sycl.hpp>
+#include <sycl/detail/core.hpp>
```
Comment (victor-eds) [inline sycl/test-e2e/PrivateAlloca/ValidUsage/Inputs/private_alloca_test.hpp:5]: Also did this in all the tests

File: sycl/test-e2e/PrivateAlloca/ValidUsage/private_alloca_bool_size.cpp [hunk 4]
```diff
@@ -1,6 +1,5 @@
 // RUN: %{build} -w -o %t.out
 // RUN: echo 1  | %{run} %t.out
-// UNSUPPORTED: cuda || hip
```
Comment (victor-eds) [inline sycl/test-e2e/PrivateAlloca/ValidUsage/private_alloca_bool_size.cpp:3]: Not needed, using `lit.local.cfg`

File: sycl/test/optional_kernel_features/private_alloca.cpp [hunk 5]
```diff
@@ -0,0 +1,31 @@
+// RUN: %clangxx %s -fsycl -fsycl-device-only -S -emit-llvm -o - | FileCheck %s
```
Comment (victor-eds) [inline sycl/test/optional_kernel_features/private_alloca.cpp:1]: Run LLVM passes to check propagation of used aspects

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 6]
```diff
@@ -23817,9 +23817,25 @@ CodeGenFunction::EmitIntelSYCLAllocaBuiltin(const CallExpr *E,
...
+    llvm::Function *F = CI->getCalledFunction();
+    constexpr llvm::StringLiteral MDName = "sycl_used_aspects";
+    if (!F->getMetadata(MDName)) {
+      auto *AspectAttr = FD->getAttr<SYCLUsesAspectsAttr>();
+      assert(AspectAttr && AspectAttr->aspects_size() == 1 &&
+             "Expecting a single aspect");
+      llvm::APSInt AspectInt =
+          (*AspectAttr->aspects_begin())->EvaluateKnownConstInt(getContext());
+      llvm::Constant *C =
+          Builder.getInt32(static_cast<int32_t>(AspectInt.getZExtValue()));
```
Comment (premanandrao) [inline clang/lib/CodeGen/CGBuiltin.cpp:23834]: Does this need to be SExt instead of ZExt?

File: clang/test/CodeGenSYCL/builtin-alloca.cpp [hunk 7]
```diff
@@ -46,3 +46,12 @@ SYCL_EXTERNAL void test(sycl::kernel_handler &kh) {
...
 }
+
+// CHECK: declare !sycl_used_aspects ![[#USED_ASPECTS:]] ptr @llvm.sycl.alloca.p0.p4.p4.p4.f64
+
+// CHECK: declare !sycl_used_aspects ![[#USED_ASPECTS]] ptr @llvm.sycl.alloca.p0.p4.p4.p4.i32
+
+// CHECK: declare !sycl_used_aspects ![[#USED_ASPECTS]] ptr @llvm.sycl.alloca.p0.p4.p4.p4.s_struct.myStructs
+
+// CHECK-DAG: ![[#USED_ASPECTS]] = !{i32 [[#PRIVATE_ALLOCA_ASPECT:]]}
+// CHECK-DAG: !{!"ext_oneapi_private_alloca", i32 [[#PRIVATE_ALLOCA_ASPECT]]}
```
Comment (premanandrao) [inline clang/test/CodeGenSYCL/builtin-alloca.cpp:57]: For my curiosity, why are these CHECK-DAGs?

Comment (victor-eds) [reply clang/test/CodeGenSYCL/builtin-alloca.cpp:57]: AFAIK, we could have them in reverse order.

File: clang/lib/CodeGen/CGBuiltin.cpp (hunk 6 above)
Comment (victor-eds) [reply clang/lib/CodeGen/CGBuiltin.cpp:23834]: I changed to `getIntegerValue` :+1:

**Review by premanandrao — APPROVED** [review]
FE changes LGTM

**Comment by victor-eds** [comment]
@intel/dpcpp-tools-reviewers , @intel/llvm-reviewers-runtime, can I have a quick look here, please?

File: llvm/tools/sycl-post-link/SpecConstants.cpp [hunk 8]
```diff
@@ -982,6 +969,21 @@ PreservedAnalyses SpecConstantsPass::run(Module &M,
             generateSpecConstDefaultValueMetadata(DefaultValue));
       }
 
+      if (IsSYCLAlloca) {
+        // In case this is a 'sycl.llvm.alloca' intrinsic, use the emitted
+        // specialization constant as the allocation size.
+        auto *Intr = cast<SYCLAllocaInst>(CI);
+        // For emulation mode, use the default value for now. This code should
+        // never be run, as the runtime should throw a 'kernel_not_supported'
+        // exception.
```
Comment (againull) [inline llvm/tools/sycl-post-link/SpecConstants.cpp:978]: From my understanding if we have Level Zero/OpenCL AOT compilation scenario, then  we use emulation mode (HandlingMode::emulation) and in general case user may set new value for spec constant representing the size of alloca. As implemented in this PR - the new aspect is considered supported for Level Zero/OpenCL.
Could you please clarify where do  we throw kernel_not_supported error in such case? I.e. I don't understand where we differentiate between SPIRV/AOT case to throw such error?

Comment (victor-eds) [reply llvm/tools/sycl-post-link/SpecConstants.cpp:985]: True. We should fail in that case. AOT compilation does not match with what we wanna achieve here. I added a check for emulated mode + SPIR-V target.

I will add an amendment to the spec and handle that in the frontend to report the error as soon as possible. Thanks!

**Review by AlexeySachkov — APPROVED** [review]

**Review by smanna12 — APPROVED** [review]
FE changes LGTM

**Comment by victor-eds** [comment]
@intel/llvm-gatekeepers can we get this one merged? CI errors look unrelated to this change.

**Comment by againull** [comment]
********************
Failed Tests (1):
  SYCL :: Basic/multi_ptr_null_relational_operators.cpp

failure is unrelated and fixed in https://github.com/intel/llvm/pull/13268

**Pull request merged**